design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,284 @@
+```
+
+#### Considered Alternative
+
+It is possible to use a meta directive (see below) in order to drive asserts as well. This was
+considered and rejected because it would require additional parsing and more testing-specific code
+to convert to expressions and pass into the codegen, which diverges from the motivating principle 
+of minimizing testing-only code.
+
+### Meta Directives
+
+Some functionality cannot be addressed in standard SQL syntax and will instead be supported using
+meta directives. Some examples of these directives:
+
+- The `--@expected.error` and `--@expected.message` directives will indicate that the below test 
+    expects an exception to be thrown with the specified message and type.
+- The `--@topic.denylist` directive will ensure that any topics matching the denylist will not
+    be present at the end of the test execution.
+
+### Global Checks
+
+These are a set of checks that will happen without any directive or `ASSERT` statement. Some
+checks that we may consider including:
+
+- _Processing Log Check_: this check will ensure that there are no failures in the processing log.
+    If the test case intends to check that the processing log contains certain entries, this check
+    can be disabled on a test-by-test basis and assert via `ASSERT DATA` on the processing log
+    stream.
+    
+### Language Gaps
+
+There are some features that we need to implement in ksqlDB's SQL language in order to put YATT
+on par with the existing testing tools. Some of those are outlined here:
+
+- _Windowed Keys_: when we have structured key support, we can `ASSERT DATA` and just provide the
+    struct as the value. Until then, windowed keys will either be unsupported or stringified.
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and YATT.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
+- _Other Unsupported Inserts_: there are other types of inserts that ksqlDB doesn't allow for at the
+    moment (e.g. `enum` support or binary formats). This is only somewhat a regression from the 
+    existing (R)QTT tests, so we may or may not decide to support it in YATT.
+    
+### Error Reporting & UX
+
+One of the bigger concerns with the existing ksql-test-runner is that it's helpful in letting
+us know _that_ an error happened, but not so much _why_ it happened or even at times _what_
+exactly happened. To avoid this pitfall, the errors will include:
+
+- the assert statement that failed, with a link to the line that failed in the test file
+- the actual data, and the expected data
+- the ability to add `PRINT` statements to debug the issue further by printing topic contents
```

Comment:
Does this mean repurposing the existing print topic syntax for use in the testing tool? This functionality makes a lot of sense but I wonder if repurposing existing syntax for something else will become confusing or problematic if we later add support for push queries/pull queries/print topic statements in the testing tool.

Reply:
I'm not entirely sure I understand the concern?

Reply:
My original thinking was: If YATT is eventually extended to include support for the current RQTT, then YATT will have to support statements for push and pull queries. If it's eventually extended to also support `PRINT <topic>;` statements (a natural extension to push query support), but the `PRINT <topic>;` syntax already has a specific meaning for the testing tool (i.e., print contents of the topic from the beginning) which conflicts with the meaning of the KSQL statement (print newly arriving data only), then it'll be confusing what the expected behavior is. To remedy the confusion we could:
- have YATT assume `auto.offset.reset=earliest` by default, so the KSQL statement `PRINT <topic>;` also prints from the beginning
- require KSQL syntax for printing topics in YATT, e.g., `PRINT 'foo';` complete with single quotes and everything. (I assume this was already the plan, in order to reuse the parser.)

What are your thoughts on the former (defaulting to `auto.offset.reset=earliest`)? While reading the proposal I had in my mind we wouldn't do that, though the KLIP wasn't explicit one way or the other.

Reply:
ah, my intention was actually to have YATT's `PRINT` statement match the semantics of KSQL's `PRINT` statement - the only difference would be that it knows when the topic ends and it doesn't hang forever (to be automation friendly) and instead ends on topic end. 

That means that if you wanted to print from beginning you would need `PRINT 'foo' FROM BEGINNING;`.

Reply:
Got'cha, that makes sense to me then. Thanks!